engine/server/server.go

```diff
@@ -164,291 +240,228 @@ func (srv *DaggerServer) ServeClientConn(
-			c.readCond.L.Unlock()
-			// return early without calling inner Read
-			return 0, os.ErrDeadlineExceeded
-		}
-
-		go func() {
-			dt := time.Until(c.readDeadline)
-			if dt > 0 {
-				time.Sleep(dt)
-			}
-
-			cancel()
-		}()
-	}
-
-	// Keep track of the reader so a future SetReadDeadline can interrupt it.
-	c.readers = append(c.readers, cancel)
-
-	c.readCond.L.Unlock()
-
-	// Start a goroutine for the actual Read operation
-	read := make(chan struct{})
-	var rN int
-	var rerr error
-	go func() {
-		defer close(read)
-
-		c.readCond.L.Lock()
-		defer c.readCond.L.Unlock()
-
-		for ctx.Err() == nil {
-			if c.readEOF {
-				rerr = io.EOF
-				break
+	srv := handler.NewDefaultServer(schema)
+	// NB: break glass when needed:
+	// srv.AroundResponses(func(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	// 	res := next(ctx)
+	// 	pl, err := json.Marshal(res)
+	// 	slog.Debug("graphql response", "response", string(pl), "error", err)
+	// 	return res
+	// })
+	mux := http.NewServeMux()
+	mux.Handle("/query", srv)
+	mux.Handle("/shutdown", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		bklog.G(ctx).Debugf("shutting down client %s", clientMetadata.ClientID)
+		defer bklog.G(ctx).Debugf("shut down client %s", clientMetadata.ClientID)
+
+		if len(s.upstreamCacheExporterCfgs) > 0 && clientMetadata.ClientID == s.mainClientCallerID {
+			bklog.G(ctx).Debugf("running cache export for client %s", clientMetadata.ClientID)
+			cacheExporterFuncs := make([]buildkit.ResolveCacheExporterFunc, len(s.upstreamCacheExporterCfgs))
+			for i, cacheExportCfg := range s.upstreamCacheExporterCfgs {
+				cacheExportCfg := cacheExportCfg
+				cacheExporterFuncs[i] = func(ctx context.Context, sessionGroup session.Group) (remotecache.Exporter, error) {
+					exporterFunc, ok := s.upstreamCacheExporters[cacheExportCfg.Type]
+					if !ok {
+						return nil, fmt.Errorf("unknown cache exporter type %q", cacheExportCfg.Type)
+					}
+					return exporterFunc(ctx, sessionGroup, cacheExportCfg.Attrs)
+				}
 			}
-
-			n, _ := c.readBuf.Read(b) // ignore EOF here
-			if n > 0 {
-				rN = n
-				break
+			s.clientCallMu.RLock()
+			bk := s.clientCallContext[""].Root.Buildkit
+			s.clientCallMu.RUnlock()
+			err := bk.UpstreamCacheExport(ctx, cacheExporterFuncs)
+			if err != nil {
+				bklog.G(ctx).WithError(err).Errorf("error running cache export for client %s", clientMetadata.ClientID)
 			}
-
-			c.readCond.Wait()
+			bklog.G(ctx).Debugf("done running cache export for client %s", clientMetadata.ClientID)
 		}
-	}()
 
-	// Wait for either Read to complete or the timeout
-	select {
-	case <-read:
-		return rN, rerr
-	case <-ctx.Done():
-		return 0, os.ErrDeadlineExceeded
+		// TODO:(XXX) double-check this is still correct, don't break services across modules
+		if err := s.services.StopClientServices(ctx, clientMetadata); err != nil {
+			bklog.G(ctx).WithError(err).Error("failed to shutdown")
+		}
```

Comment:
Ah, unrelated to this PR in general, but this shouldn't be in `/shutdown`. If a client doesn't call this (by accident, most likely by crashing), then the services keep running indefinitely.

Is there any way we could tie this to the lifetime of `ServeClientConn` instead, and always call it at the end of that?

I think there are two behaviors we want for "shutdown":
- Cleanup of internal resources, we always want to do this, like shutting down services
- Cleanup of the session controlled by the client, like upstream cache exports

Reply:
Oh sure, yeah I can just move `StopClientServices` to `DaggerServer.Close`, which should result in it always getting called (since `Close` is called from the `BuildkitController` which is tied to the lifetime of the whole process). Very simple so will just do that here even though it's slightly orthogonal to the other changes.

Reply:
Yeah that should do it :tada: Thanks :pray:

Reply:
Just for continuity, sounds like that will effectively implement the second part of this fix in https://github.com/dagger/dagger/issues/6493#issuecomment-1918679919 right?

Reply:
I think this more just results in the current behavior happening consistently, even if the client exits ungracefully and never calls `/shutdown`. It still always force stops for every client (whereas that comment is suggesting changing the behavior such that force stop is only on main client, detach for others).